lib/Arch/SPARC32_C.cpp

```diff
@@ -0,0 +1,347 @@
+  auto i32_ty = llvm::Type::getInt32Ty(type->getContext());
+  return llvm::ArrayType::get(i32_ty, num_elements);
+}
+
+}  // namespace
+
+// This is the only calling convention for 32-bit SPARC code.
+class SPARC32_C : public CallingConvention {
+ public:
+   explicit SPARC32_C(const remill::Arch *arch);
+   virtual ~SPARC32_C(void) = default;
+
+   llvm::Error AllocateSignature(FunctionDecl &fdecl,
+                                 llvm::Function &func) override;
+
+  private:
+   llvm::Error BindParameters(llvm::Function &function, bool injected_sret,
+                              std::vector<ParameterDecl> &param_decls);
+
+   llvm::Error BindReturnValues(llvm::Function &function,
+                                bool &injected_sret,
+                                std::vector<ValueDecl> &ret_decls);
+
+   const std::vector<RegisterConstraint> &parameter_register_constraints;
+   const std::vector<RegisterConstraint> &return_register_constraints;
+};
+
+std::unique_ptr<CallingConvention> CallingConvention::CreateSPARC32_C(
+      const remill::Arch *arch) {
+  return std::unique_ptr<CallingConvention>(new SPARC32_C(arch));
+}
+
+SPARC32_C::SPARC32_C(const remill::Arch *arch)
+    : CallingConvention(llvm::CallingConv::C, arch),
+      parameter_register_constraints(kParamRegConstraints),
+      return_register_constraints(kReturnRegConstraints) {}
+
+// Allocates the elements of the function signature of func to memory or
+// registers. This includes parameters/arguments, return values, and the return
+// stack pointer.
+llvm::Error SPARC32_C::AllocateSignature(FunctionDecl &fdecl,
+                                         llvm::Function &func) {
+  // Bind return values first to see if we have injected an sret into the
+  // parameter list. Then, bind the parameters. It is important that we bind the
+  // return values before the parameters in case we inject an sret.
+  bool injected_sret = false;
+  auto err = BindReturnValues(func, injected_sret, fdecl.returns);
+  if (remill::IsError(err)) {
+    return err;
+  }
+
+  err = BindParameters(func, injected_sret, fdecl.params);
+  if (remill::IsError(err)) {
+    return err;
+  }
+
+  fdecl.return_stack_pointer_offset = 0;
+  fdecl.return_stack_pointer = arch->RegisterByName("o6");
+
+  fdecl.return_address.reg = arch->RegisterByName("o7");
+  fdecl.return_address.type = fdecl.return_address.reg->type;
+
+  return llvm::Error::success();
+}
+
+llvm::Error SPARC32_C::BindReturnValues(
+    llvm::Function &function, bool &injected_sret,
+    std::vector<anvill::ValueDecl> &ret_values) {
+
+  llvm::Type *ret_type = function.getReturnType();
+  injected_sret = false;
+
+  // If there is an sret parameter then it is a special case. For the X86_C ABI,
+  // the sret parameters are guarenteed the be in %eax. In this case, we can
+  // assume the actual return value of the function will be the sret struct
+  // pointer.
+  if (function.hasStructRetAttr()) {
+    ret_values.emplace_back();
```

Comment:
So in C++17 (which I think we have on? I can't confirm if we still need C++14 support) the emplace_back() call should return a reference to the newly created element. If we enforce 17, then we can get rid of all the .back() calls.